src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp

```diff
@@ -443,10 +444,24 @@ bool PDPClient::create_ds_pdp_reliable_endpoints(
     {
         eprosima::shared_lock<eprosima::shared_mutex> disc_lock(mp_builtin->getDiscoveryMutex());
 
+        bool set_logicals = mp_RTPSParticipant->has_tcp_transports();
+
         for (const eprosima::fastdds::rtps::RemoteServerAttributes& it : mp_builtin->m_DiscoveryServers)
         {
-            mp_RTPSParticipant->createSenderResources(it.metatrafficMulticastLocatorList);
-            mp_RTPSParticipant->createSenderResources(it.metatrafficUnicastLocatorList);
+            if (set_logicals)
+            {
+                LocatorSelectorEntry entry(pattr.allocation.locators.max_unicast_locators,
+                        pattr.allocation.locators.max_multicast_locators);
+                entry.is_initial_peer_or_ds = true;
+                entry.fill_multicast(it.metatrafficMulticastLocatorList);
+                entry.fill_unicast(it.metatrafficUnicastLocatorList);
+                mp_RTPSParticipant->createSenderResources(entry);
+            }
+            else
+            {
+                mp_RTPSParticipant->createSenderResources(it.metatrafficMulticastLocatorList);
+                mp_RTPSParticipant->createSenderResources(it.metatrafficUnicastLocatorList);
+            }
```

Comment:
I think we can always follow the new approach (note the usage of the builder method in my suggestion):

```suggestion
            {
                auto entry = LocatorSelectorEntry::create_fully_selected_entry(
                        it.metatrafficMulticastLocatorList, it.metatrafficUnicastLocatorList);
                mp_RTPSParticipant->createSenderResources(entry);
            }
```

Reply:
Applies elsewhere